xdsl/dialects/builtin.py

```diff
@@ -1436,7 +1457,7 @@ def print(self, printer: Printer) -> None:
 f128 = Float128Type()
 
 
-_MemRefTypeElement = TypeVar("_MemRefTypeElement", bound=Attribute)
+_MemRefTypeElement = TypeVar("_MemRefTypeElement", bound=Attribute, covariant=True)
```

Comment:
this change was made to appease Pyright below, but I'm not really sure that it's kosher... My understanding is that, even though a memref is mutable, and hence a memref type parameter of the memref itself should be covariant, a memref constraint is immutable, so a memref constraint that only accepts cats should be ok to give whenever the client expects a constraint that only accepts animals. Does this seem right? Am I missing something?

Reply:
If you can store stuff in the memref then this is technically unsound I believe

Reply:
Actually, I can't even really figure out why this type is generic at all?

Reply:
This is the type not the memref itself. I am now pretty convinced that it's sound, just because all attributes and attribute constraints are immutable, so it makes sense to have their type parameters be covariant. It is not really related to the type of the objects in the IR, as far as the Python type system is concerned.

Reply:
Looking through the uses of `MemRefType`, I'm pretty sure this generic argument serves no function whatsoever